pkg/controller/zerocapacity/zero_controller.go

```diff
@@ -0,0 +1,448 @@
+	// ZeroInitialized means that all required resources have been initialized
+	ZeroInitialized Phase = "Initialized"
+	// ZeroRunning means that the required action has been initiated on the infinispan server.
+	ZeroRunning Phase = "Running"
+	// ZeroSucceeded means that the action on the server has completed and the zero pod has been terminated.
+	ZeroSucceeded Phase = "Succeeded"
+	// ZeroFailed means that the action failed on the infinispan server and the zero pod has terminated.
+	ZeroFailed Phase = "Failed"
+	// ZeroUnknown means that for some reason the state of the action could not be obtained, typically due
+	// to an error in communicating with the underlying zero pod.
+	ZeroUnknown Phase = "Unknown"
+)
+
+func CreateController(name string, reconciler Reconciler, mgr manager.Manager) error {
+	r := &Controller{
+		Name:       name,
+		Client:     mgr.GetClient(),
+		Reconciler: reconciler,
+		Kube:       kube.NewKubernetesFromController(mgr),
+		Log:        logf.Log.WithName(name),
+		Scheme:     mgr.GetScheme(),
+	}
+
+	// Create a new controller
+	c, err := controller.New(name, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource
+	err = c.Watch(&source.Kind{Type: reconciler.Type()}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return c.Watch(&source.Kind{Type: &corev1.Pod{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    reconciler.Type(),
+	})
+}
+
+func (z *Controller) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reconciler := z.Reconciler
+	resource := reflect.TypeOf(reconciler.Type()).Name()
+	namespace := request.Namespace
+
+	reqLogger := z.Log.WithValues("Request.Namespace", namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling " + resource)
+	defer reqLogger.Info("----- End Reconciling " + resource)
+
+	instance, err := reconciler.ResourceInstance(request.NamespacedName, z)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch %s CR '%s': %w", resource, request.Name, err)
+	}
+
+	phase := instance.Phase()
+	switch phase {
+	case "":
+		return reconcile.Result{}, instance.UpdatePhase(ZeroInitializing)
+	case ZeroInitializing:
+		return z.initializeResources(request, instance)
+	case ZeroSucceeded, ZeroFailed:
+		return z.cleanupResources(request)
+	}
+
+	infinispan := &v1.Infinispan{}
+	clusterName := instance.Cluster()
+	clusterObjKey := client.ObjectKey{
+		Namespace: namespace,
+		Name:      clusterName,
+	}
+	if err := z.Get(context.Background(), clusterObjKey, infinispan); err != nil {
```

Comment:
@ryanemerson what do you think about option backup to be deleted when delete Infinispan CR?

Reply:
I think this is something we should maybe consider adding later. Personally I think it's maybe an anti-pattern, as you normally want the backup to restore content when a Infinispan cluster has been recreated, so deleting Backup content on Infinispan CR removal could cause confusion.